test-interface/src/main/scala/org/scalajs/testinterface/internal/Slave.scala

```diff
@@ -5,119 +5,70 @@ import js.annotation._
 
 import sbt.testing._
 
-import scala.collection.mutable
+import scala.concurrent.{Future, Promise}
+import scala.concurrent.duration._
+import scala.concurrent.ExecutionContext.Implicits.global
 
 import scala.util.control.NonFatal
-import scala.util.{Try, Success, Failure}
+import scala.util.Try
 
+import org.scalajs.testcommon._
 import org.scalajs.testinterface.ScalaJSClassLoader
 
 @JSExportTopLevel("org.scalajs.testinterface.internal.Slave")
 final class Slave(frameworkName: String, args: js.Array[String],
-    remoteArgs: js.Array[String]) extends BridgeBase(frameworkName) {
-
-  // State
-
-  private[this] var canSendRunnerMessage = false
-
-  /** Queue for messages from the slave runner to the master runner */
-  private[this] val messageQueue = mutable.Queue.empty[String]
+    remoteArgs: js.Array[String]) {
 
   private[this] var runner: Runner = _
 
-  protected def handleMsgImpl(cmd: String, strArg: => String): Unit = {
-    def jsonArg = js.JSON.parse(strArg)
-    allowSendRunnerMessage {
-      cmd match {
-        case "newRunner" =>
-          reply(newRunner())
-        case "execute" =>
-          // No reply here. execute is async
-          execute(jsonArg)
-        case "stopSlave" =>
-          reply(stopSlave())
-        case "msg" =>
-          val res = incomingRunnerMessage(strArg)
-          // Only reply if something failed
-          if (res.isFailure)
-            reply(res)
-        case cmd =>
-          throw new IllegalArgumentException(s"Unknown command: $cmd")
-      }
-    }
-  }
-
-  // Runner message handler methods
-
-  private def outboundRunnerMessage(msg: String): Unit =
-    if (canSendRunnerMessage) sendOutboundRunnerMessage(msg)
-    else messageQueue.enqueue(msg)
-
-  private def sendOutboundRunnerMessage(msg: String): Unit = {
-    assert(canSendRunnerMessage)
-    Com.send(s"msg:$msg")
-  }
-
-  private def allowSendRunnerMessage[T](body: => T): T = {
-    try {
-      canSendRunnerMessage = true
-
-      // Flush the queue
-      while (!messageQueue.isEmpty)
-        sendOutboundRunnerMessage(messageQueue.dequeue)
-
-      body
-    } finally {
-      canSendRunnerMessage = false
-    }
-  }
+  JSRPC.attach(JSSlaveEndpoints.newRunner)(newRunner _)
+  JSRPC.attachAsync(JSSlaveEndpoints.execute)(execute _)
+  JSRPC.attach(JSSlaveEndpoints.stopSlave)(stopSlave _)
+  JSRPC.attach(JSSlaveEndpoints.msg)(runner.receiveMessage _)
```

Comment:
It "NPE"s here. `runner` is already set to `null`, but it still get's a message.